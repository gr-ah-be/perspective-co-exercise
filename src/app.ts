import express, { Express, Request, Response } from 'express';
import { userRouter } from './routers/user.router';
import * as dotenv from 'dotenv';
import cors from 'cors';
import pinoHttp from 'pino-http';
import pino from 'pino';
import { errorHandler } from './routers/middlewares/error.middleware';

dotenv.config();

export const buildApplication = (logger: pino.Logger | typeof console): express.Express => {
    const app: Express = express();

    app.use(pinoHttp({ logger: logger as pino.Logger }));
    app.use(cors()).use(express.json()).options('*', cors());

    app.use('/users', userRouter);

    app.use(errorHandler);

    app.get('/health', (_req: Request, res: Response) => {
        res.status(200).send({ status: 'UP' });
    });

    return app;
};
