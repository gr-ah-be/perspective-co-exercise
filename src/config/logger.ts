import pino from 'pino';

const LOG_LEVEL = process.env.LOG_LEVEL || 'info';

export const logger = pino({ level: LOG_LEVEL });
