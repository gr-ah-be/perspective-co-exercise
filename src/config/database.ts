import mongoose from 'mongoose';
import dotenv from 'dotenv';
import { logger } from './logger';

dotenv.config();

const mongoUri = process.env.MONGO_URI || 'mongodb://localhost:27017/backend-test';

export const connectDB = async (connectionUrl?: string) => {
    try {
        await mongoose.connect(connectionUrl ?? mongoUri);
        logger.info('Database connection established successfully.');
        mongoose.connection.on('disconnected', () => {
            logger.info('Disconnected from database');
        });
    } catch (error) {
        logger.error('Failed to connect to MongoDB:', error);
    }
};

export const disconnectDB = async () => {
    try {
        await mongoose.disconnect();
        logger.info('Disconnected from database');
    } catch (error) {
        logger.error('Failed to disconnect from MongoDB:', error);
    }
};
