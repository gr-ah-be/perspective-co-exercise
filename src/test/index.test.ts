import { afterAll, afterEach, beforeAll, describe, expect, it } from '@jest/globals';
import request from 'supertest';
import { buildApplication } from '../app';
import { logger } from '../config/logger';
import * as http from 'http';
import { setup, destroy, clearData, insertData } from './database-test';
let app;
let server: http.Server;

describe.only('User API Endpoints', () => {
    beforeAll(async () => {
        app = buildApplication(logger);
        server = app.listen(3000);
        await setup();
    });

    afterAll(async () => {
        await destroy();
        await server.close();
    });

    describe('POST /users', () => {
        afterEach(async () => {
            await clearData();
        });

        it('should create a new user successfully', async () => {
            const response = await request(server).post('/users').send({
                firstName: 'John',
                lastName: 'Doe',
                email: 'john.doe@example.com',
                phone: '+491516644904',
            });
            expect(response.status).toBe(201);
            expect(response.body).toHaveProperty('_id');
            expect(response.body.email).toBe('john.doe@example.com');
        });

        it('should return validation error for missing fields', async () => {
            const response = await request(server).post('/users').send({});
            expect(response.status).toBe(400);
        });

        it('should fail when email format is invalid', async () => {
            const response = await request(server).post('/users').send({
                firstName: 'John',
                lastName: 'Doe',
                email: 'invalid-email',
                phone: '1234567890',
            });
            expect(response.status).toBe(400);
            expect(response.body).toMatchObject({
                errors: [
                    {
                        validation: 'email',
                        code: 'invalid_string',
                        message: 'Invalid email format',
                        path: ['email'],
                    },
                ],
            });
        });

        it('should return error when creating a user with a duplicate email', async () => {
            const response = await request(server).post('/users').send({
                firstName: 'John',
                lastName: 'Doe',
                email: 'john.doe@example.com',
                phone: '+491516644904',
            });

            expect(response.status).toBe(201);

            // Attempt to create another user with the same email
            const duplicateResponse = await request(server).post('/users').send({
                firstName: 'Jane',
                lastName: 'Doe',
                email: 'john.doe@example.com',
                phone: '+331516644902',
            });

            expect(duplicateResponse.status).toBe(400);
            expect(duplicateResponse.body).toMatchObject({
                message: 'User with provided email already exists',
                errors: [
                    {
                        field: 'email',
                        value: 'john.doe@example.com',
                        message: 'User with provided email already exists',
                        code: 'DUPLICATE',
                    },
                ],
            });
        });

        it('should return error when creating a user with a duplicate phone', async () => {
            const response = await request(server).post('/users').send({
                firstName: 'John',
                lastName: 'Doe',
                email: 'john.doe@example.com',
                phone: '+491516644904',
            });

            expect(response.status).toBe(201);

            // Attempt to create another user with the same phone
            const duplicateResponse = await request(server).post('/users').send({
                firstName: 'Jane',
                lastName: 'Doe',
                email: 'jane.doe@example.com',
                phone: '+491516644904',
            });

            expect(duplicateResponse.status).toBe(400);
            expect(duplicateResponse.body).toMatchObject({
                message: 'User with provided phone already exists',
                errors: [
                    {
                        field: 'phone',
                        value: '+491516644904',
                        message: 'User with provided phone already exists',
                        code: 'DUPLICATE',
                    },
                ],
            });
        });
    });

    describe('GET /users', () => {
        beforeAll(async () => {
            await insertData();
        });

        it('should retrieve all users with default pagination', async () => {
            const response = await request(server).get('/users');
            expect(response.status).toBe(200);
            expect(Array.isArray(response.body)).toBe(true);
        });

        it('should retrieve users sorted by creation date descending', async () => {
            const response = await request(server).get('/users?created=desc');
            expect(response.status).toBe(200);
            expect(response.body).toBeDefined();
            expect(response.body).toMatchSnapshot();
        });

        it('should retrieve users sorted by creation date ascending', async () => {
            const response = await request(server).get('/users?created=asc');
            expect(response.status).toBe(200);
            expect(response.body).toBeDefined();
            expect(response.body).toMatchSnapshot();
        });
    });
});
