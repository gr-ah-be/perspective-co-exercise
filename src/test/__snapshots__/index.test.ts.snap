// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User API Endpoints GET /users should retrieve users sorted by creation date ascending 1`] = `
[
  {
    "_id": "674211b69b10137e588d07ec",
    "createdAt": "2018-02-25T00:15:04.327Z",
    "email": "angela.luettgen@example.com",
    "firstName": "Angela",
    "lastName": "Luettgen",
    "phone": "+497263931334",
    "updatedAt": "2018-02-25T00:15:04.327Z",
  },
  {
    "_id": "674211b69b10137e588d07f3",
    "createdAt": "2019-02-18T15:40:54.327Z",
    "email": "isabella.martinez@example.com",
    "firstName": "Isabella",
    "lastName": "Martinez",
    "phone": "+497822143294",
    "updatedAt": "2019-02-18T15:40:54.327Z",
  },
  {
    "_id": "674211b69b10137e588d07ea",
    "createdAt": "2020-05-29T04:11:00.469Z",
    "email": "johnny.kris@example.com",
    "firstName": "Johnny",
    "lastName": "Kris",
    "phone": "+493965439731",
    "updatedAt": "2020-05-29T04:11:00.469Z",
  },
  {
    "_id": "674211b69b10137e588d07e9",
    "createdAt": "2020-07-04T17:00:39.515Z",
    "email": "kasandra.pagac@example.com",
    "firstName": "Kasandra",
    "lastName": "Pagac",
    "phone": "+492594137122",
    "updatedAt": "2020-07-04T17:00:39.515Z",
  },
  {
    "_id": "674211b69b10137e588d07f2",
    "createdAt": "2020-09-21T07:15:24.327Z",
    "email": "lucas.rodriguez@example.com",
    "firstName": "Lucas",
    "lastName": "Rodriguez",
    "phone": "+499824017728",
    "updatedAt": "2020-09-21T07:15:24.327Z",
  },
  {
    "_id": "674211b69b10137e588d07e5",
    "createdAt": "2020-12-15T18:45:43.068Z",
    "email": "carmine.von@example.com",
    "firstName": "Carmine",
    "lastName": "Von",
    "phone": "+491787079926",
    "updatedAt": "2020-12-15T18:45:43.068Z",
  },
  {
    "_id": "674211b69b10137e588d07e8",
    "createdAt": "2021-03-14T16:56:29.915Z",
    "email": "dylan.mcdermott@example.com",
    "firstName": "Dylan",
    "lastName": "McDermott",
    "phone": "+495245424569",
    "updatedAt": "2021-03-14T16:56:29.915Z",
  },
  {
    "_id": "674211b69b10137e588d07f5",
    "createdAt": "2021-06-17T16:50:44.327Z",
    "email": "mia.lopez@example.com",
    "firstName": "Mia",
    "lastName": "Lopez",
    "phone": "+495108401092",
    "updatedAt": "2021-06-17T16:50:44.327Z",
  },
  {
    "_id": "674211b69b10137e588d07e3",
    "createdAt": "2021-09-23T19:20:56.138Z",
    "email": "rosa.king@example.com",
    "firstName": "Rosa",
    "lastName": "King",
    "phone": "+491516644904",
    "updatedAt": "2021-09-23T19:20:56.138Z",
  },
  {
    "_id": "674211b69b10137e588d07f0",
    "createdAt": "2021-11-09T12:40:54.327Z",
    "email": "ethan.brown@example.com",
    "firstName": "Ethan",
    "lastName": "Brown",
    "phone": "+494569138220",
    "updatedAt": "2021-11-09T12:40:54.327Z",
  },
]
`;

exports[`User API Endpoints GET /users should retrieve users sorted by creation date descending 1`] = `
[
  {
    "_id": "674211b69b10137e588d07e6",
    "createdAt": "2024-01-05T18:27:03.667Z",
    "email": "mervin.halvorson@example.com",
    "firstName": "Mervin",
    "lastName": "Halvorson",
    "phone": "+493246425099",
    "updatedAt": "2024-01-05T18:27:03.667Z",
  },
  {
    "_id": "674211b69b10137e588d07f1",
    "createdAt": "2023-05-30T09:20:14.327Z",
    "email": "sophia.garcia@example.com",
    "firstName": "Sophia",
    "lastName": "Garcia",
    "phone": "+496374811198",
    "updatedAt": "2023-05-30T09:20:14.327Z",
  },
  {
    "_id": "674211b69b10137e588d07f6",
    "createdAt": "2023-04-09T10:05:14.327Z",
    "email": "liam.clark@example.com",
    "firstName": "Liam",
    "lastName": "Clark",
    "phone": "+498534214391",
    "updatedAt": "2023-04-09T10:05:14.327Z",
  },
  {
    "_id": "674211b69b10137e588d07ee",
    "createdAt": "2023-03-22T14:02:03.327Z",
    "email": "maxwell.smith@example.com",
    "firstName": "Maxwell",
    "lastName": "Smith",
    "phone": "+499034728209",
    "updatedAt": "2023-03-22T14:02:03.327Z",
  },
  {
    "_id": "674211b69b10137e588d07e7",
    "createdAt": "2023-01-20T06:19:54.933Z",
    "email": "haskell.lesch@example.com",
    "firstName": "Haskell",
    "lastName": "Lesch",
    "phone": "+496392900348",
    "updatedAt": "2023-01-20T06:19:54.933Z",
  },
  {
    "_id": "674211b69b10137e588d07f4",
    "createdAt": "2022-10-05T22:30:34.327Z",
    "email": "aiden.hernandez@example.com",
    "firstName": "Aiden",
    "lastName": "Hernandez",
    "phone": "+494632950153",
    "updatedAt": "2022-10-05T22:30:34.327Z",
  },
  {
    "_id": "674211b69b10137e588d07e4",
    "createdAt": "2022-09-28T16:16:46.185Z",
    "email": "samir.rice@example.com",
    "firstName": "Samir",
    "lastName": "Rice",
    "phone": "+4916014501635",
    "updatedAt": "2022-09-28T16:16:46.185Z",
  },
  {
    "_id": "674211b69b10137e588d07ef",
    "createdAt": "2022-07-12T18:00:24.327Z",
    "email": "olivia.thompson@example.com",
    "firstName": "Olivia",
    "lastName": "Thompson",
    "phone": "+492345849133",
    "updatedAt": "2022-07-12T18:00:24.327Z",
  },
  {
    "_id": "674211b69b10137e588d07eb",
    "createdAt": "2022-06-20T07:01:55.040Z",
    "email": "chance.langosh@example.com",
    "firstName": "Chance",
    "lastName": "Langosh",
    "phone": "+497186844543",
    "updatedAt": "2022-06-20T07:01:55.040Z",
  },
  {
    "_id": "674211b69b10137e588d07ed",
    "createdAt": "2022-01-14T08:00:54.327Z",
    "email": "ella.brock@example.com",
    "firstName": "Ella",
    "lastName": "Brock",
    "phone": "+495164731932",
    "updatedAt": "2022-01-14T08:00:54.327Z",
  },
]
`;
